piksi_tools/console/tracking_view.py

```diff
@@ -133,104 +133,123 @@ class TrackingView(CodeFiltered):
+                if code_is_glo(s.sid.code):
+                    sat = s.fcn - GLO_FCN_OFFSET
+                    self.glo_slot_dict[i] = s.sid.sat
+                else:
+                    sat = s.sid.sat
+                key = (s.sid.code, sat, i)
+                codes_that_came.append(key)
+                if s.cn0 != 0:
+                    self.CN0_dict[key].append(s.cn0 / 4.0)
+                    self.CN0_age[key] = t
+                received_code_list = getattr(self, "received_codes", [])
+                if s.sid.code not in received_code_list:
+                    received_code_list.append(s.sid.code)
+                    self.received_codes = received_code_list
+            for key, cno_array in list(self.CN0_dict.items()):
+                if key not in codes_that_came:
+                    cno_array.append(0)
+            self.clean_cn0(t)
+            if self.update_scheduled:
+                return
+            self.update_scheduled = True
         GUI.invoke_later(self.update_plot)
 
     def update_plot(self):
-        self.CN0_lock.acquire()
-        plot_labels = []
-        plots = []
-        # Update the underlying plot data from the CN0_dict for selected items
-        new_plot_data = {'t': self.time}
-        for k, cno_array in self.CN0_dict.items():
-            key = str(k)
-            # set plot data
-            if (getattr(self, 'show_{}'.format(int(k[0])), True)):
-                new_plot_data[key] = cno_array
-        self.plot_data.update_data(new_plot_data)
-        # Remove any stale plots that got removed from the dictionary
-        for each in self.plot.plots.keys():
-            if each not in [str(a) for a in self.CN0_dict.keys()] and each != 't':
-                try:
-                    self.plot.delplot(each)
-                except KeyError:
-                    pass
-        # add/remove plot as neccesary and build legend
-        for k, cno_array in self.CN0_dict.items():
-            key = str(k)
-            if (getattr(self, 'show_{}'.format(int(k[0])), True) and
-               not cno_array.count(0) == NUM_POINTS):
-                if key not in self.plot.plots.keys():
-                    pl = self.plot.plot(('t', key), type='line',
-                                        color=get_color(k), name=key)
+        with self.CN0_lock:
+            self.update_scheduled = False
+            plot_labels = []
+            plots = []
+            # Update the underlying plot data from the CN0_dict for selected items
+            new_plot_data = {'t': self.time}
+            for k, cno_array in self.CN0_dict.items():
+                key = str(k)
+                # set plot data
+                if (getattr(self, 'show_{}'.format(int(k[0])), True)):
+                    new_plot_data[key] = cno_array
+            self.plot_data.update_data(new_plot_data)
+            # Remove any stale plots that got removed from the dictionary
+            for each in list(self.plot.plots.keys()):
```

Comment:
as above, shouldn't need `list()`

Reply:
This was initially from a Python 3 compatibility tool (`2to3`) but just in case `self.plot.delplot(each)` modifies the underlying data structure behind `list(self.plot.plots.keys())` I'd like to retain the `list` call.